### PR TITLE
fix: sign tx with correct key during rotation

### PR DESCRIPTION
**Pull request by ramizhasan111**
# Pull Request

Closes: PRO-1310

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

*Please include a succinct description of the purpose and content of the PR. What problem does it solve, and how? Link issues, discussions, other PRs, and anything else that will help the reviewer.*

#### Non-Breaking changes

If this PR includes non-breaking changes, select the `non-breaking` label. On merge, CI will automatically cherry-pick the commit to a PR against the release branch.

### CODE REVIEW

**Comment by ramizhasan111** [comment]
So I tried to write a test and it looks like there is no easy way to add a test for this in the integration tests or the bouncer:

- For this test to work in the integration tests, we need a mock keymanager contract and a mock Vault contract that has partial functionality of the said contracts (parse, extract sig and verify, update key). This is because we need to simulate the fact that a tx with the wrong aggkey signature fails on the contract.

- For this test to work in the bouncer, we need an apicall to be initiated (threshold_sign_and_broadcast) between a rotation tx being initiated and the rotation tx being successfully signed. This is not easy to do from the bouncer since we cannot directly initiate a tx via the interface, we can request a swap which will take time to reach the Allbatch call initiation stage and could already be too late as the rotation tx could be successfully signed by then.

thoughts? @dandanlen @kylezs

**Comment by albert-llimos** [comment]
When I read this I felt optimistic that we would maybe be able to reproduce this in the bouncer in a reasonable way. I think I have already been able to reproduce the issue. It would of course be very time dependent, for the reasons that Ramiz mentioned, but hopefully it's not too flaky. I'll try to put something together.

**Comment by albert-llimos** [comment]
That worked. As I said, it's time sensitive but it seems to not be flaky at all and hit the issue. I'd go with this solution but of course if you prefer another type of test go with it. PR here with the description.
https://github.com/chainflip-io/chainflip-backend/pull/4799

**Comment by ramizhasan111** [comment]
> That worked. As I said, it's time sensitive but it seems to not be flaky at all and hit the issue. I'd go with this solution but of course if you prefer another type of test go with it. PR here with the description. #4799

nice! I think it should work very consistently since the order of events and the number of blocks between different events in the rotation and the number of blocks it takes from swap initiation to apicall initiation should all be constant in a stable localnet environment. In a rare case where it fails, rerunning the test should work.

**Comment by albert-llimos** [comment]
> nice! I think it should work very consistently since the order of events and the number of blocks between different events in the rotation and the number of blocks it takes from swap initiation to apicall initiation should all be constant in a stable localnet environment. In a rare case where it fails, rerunning the test should work.

I agree. And it won't be flaky in a sense that it will fail from time to time, but rather that it might not hit the bug if it's there. Right now, if this PR fixes it consistently, it should never fail.

File: state-chain/pallets/cf-threshold-signature/src/key_rotator.rs [hunk 1]
```diff
@@ -150,33 +150,28 @@ impl<T: Config<I>, I: 'static> KeyRotator for Pallet<T, I> {
...
-										} else {
-											true
-										}
-									});
-									*new_public_key
-								},
-								_ => unreachable!(
+					PendingKeyRotation::<T, I>::mutate(|pending_key_rotation| {
+						match pending_key_rotation {
+							Some(KeyRotationStatus::AwaitingActivation { request_ids, .. }) => {
```
[the remark this answers is not in the document]
Comment (ramizhasan111) [reply state-chain/pallets/cf-threshold-signature/src/key_rotator.rs]: removed

**Comment by ramizhasan111** [comment]
> Just some nits, I feel like we can make things a bit clearer: `AwaitingActivation` and `activate_new_key` are a bit vague, and there is nothing being activated, really.
> 
> `AwaitingActivation` could be `AwaitingActivationSignature`, which would be more accurate and would explain why it contains the pending signature requests.
> 
> `activate_new_key` could simply be `key_rotation_complete()` or something.

changed to ```AwaitingActivationSignatures``` and ```mark_key_rotation_complete```

**Review by dandanlen — APPROVED** [review]
Thanks!

**Pull request merged**